search-app/src/cmr/search/services/parameters/converters/shapefile.clj

```diff
@@ -0,0 +1,146 @@
+(ns cmr.search.services.parameters.converters.shapefile
+  "Contains parameter converters for shapefile parameter"
+  (:require 
+    [clojure.java.io :as io]
+    [clojure.string :as str]
+    [cmr.common.config :as cfg :refer [defconfig]]
+    [cmr.common.log :refer (debug)]
+    [cmr.common.mime-types :as mt]
+    [cmr.common-app.services.search.group-query-conditions :as gc]
+    [cmr.common-app.services.search.params :as p]
+    [cmr.common.services.errors :as errors]
+    [cmr.search.models.query :as qm]
+    [cmr.search.services.parameters.converters.geometry :as geo]
+    [cmr.common.util :as util])
+  (:import
+   (java.io BufferedInputStream File FileReader FileOutputStream FileInputStream)
+   (java.nio.file Files)
+   (java.nio.file.attribute FileAttribute)
+   (java.net URL)
+   (java.util HashMap)
+   (java.util.zip ZipFile ZipInputStream)
+   (org.apache.commons.io FilenameUtils)
+   (org.geotools.data DataStoreFinder FileDataStoreFinder Query)
+   (org.geotools.data.simple SimpleFeatureSource)
+   (org.geotools.data.geojson GeoJSONDataStore)
+   (org.geotools.geometry.jts JTS)
+   (org.geotools.referencing CRS)
+   (org.geotools.util URLs)))
+
+(def EPSG-4326-CRS (CRS/decode "EPSG:4326" true))
+
+(defconfig enable-shapefile-parameter-flag
+  "Flag that indicates if we allow spatial searching by shapefile."
+  {:default true :type Boolean})
+
+(defn- unzip-file
+  "Unzip a file (of type File) into a temporary directory and return the directory path as a File"
+  [source]
+  (try
+    (let [target-dir (Files/createTempDirectory "Shapes" (into-array FileAttribute []))]
+      (with-open [zip (ZipFile. source)]
+        (let [entries (enumeration-seq (.entries zip))
+              target-file #(File. (.toString target-dir) (str %))]
+          (doseq [entry entries :when (not (.isDirectory ^java.util.zip.ZipEntry entry))
+                  :let [f (target-file entry)]]
+            (debug (format "Zip file entry: [%s]" (.getName entry)))
+            (io/copy (.getInputStream zip entry) f))))
+      (.toFile target-dir))
+    (catch Exception e
+      (errors/throw-service-error :bad-request (str "Error while uncompressing zip file: " (.getMessage e))))))
```

Comment:
It's possible that this could be an internal server error too, but probably difficult to tell so I guess always returning a 400 is ok.